### PR TITLE
Move `reset recovery` into RecoveriesCollection

### LINKED ISSUE #19473: Cancellation of recovery deletes files still held onto by writes
The RepositoryUpgradabilityIT test fails regularly because indices are deleted before trying to recreate the index for other purposes.  We could add `ensureGreen` calls after the index deletions, but the underlying cause of failure is that the ongoing recoveries get cancelled and this causes shard level data to be deleted.  However, the deletion of these files can happen *while* other threads are holding onto these files for the purposes of writes, which trips assertions where the write operations expects the temp files to be there (but they have been deleted by the recovery cancellation).

An example stack trace:
```
ERROR   21.6s J0 | RepositoryUpgradabilityIT.testRepositoryWorksWithCrossVersions <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=11223, name=elasticsearch[node_td1][generic][T#6], state=RUNNABLE, group=TGRP-RepositoryUpgradabilityIT]
   > 	at __randomizedtesting.SeedInfo.seed([6C53C46436F84D7A:76F1740152D478D]:0)
   > Caused by: java.lang.AssertionError: expected: [recovery.1468504452288._0.cfe] in []
   > 	at __randomizedtesting.SeedInfo.seed([6C53C46436F84D7A]:0)
   > 	at org.elasticsearch.indices.recovery.RecoveryTarget.assertTempFileExists(RecoveryTarget.java:407)
   > 	at org.elasticsearch.indices.recovery.RecoveryTarget.writeFileChunk(RecoveryTarget.java:397)
   > 	at org.elasticsearch.indices.recovery.RecoveryTargetService$FileChunkTransportRequestHandler.messageReceived(RecoveryTargetService.java:417)
   > 	at org.elasticsearch.indices.recovery.RecoveryTargetService$FileChunkTransportRequestHandler.messageReceived(RecoveryTargetService.java:390)
   > 	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
   > 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69)
   > 	at org.elasticsearch.transport.local.LocalTransport$1.doRun(LocalTransport.java:322)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:510)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   > 	at java.lang.Thread.run(Thread.java:745)```

### PR DESCRIPTION
**Pull request by s1monw**
Today when we reset a recovery because of the source not being
ready or the shard is getting removed on the source (for whatever reason)
we wipe all temp files and reset the recovery without respecting any
reference counting or locking etc. all streams are closed and files are
wiped. Yet, this is problematic since we assert that some files are on disk
etc. when we finish writing a file. These assertions don't hold anymore if we
concurrently wipe the tmp files.

This change moves the logic out of RecoveryTarget into RecoveriesCollection which
basically clones the RecoveryTarget on reset instead which allows in-flight operations
to finish gracefully. This means we now have a single path for cleanups in RecoveryTarget
and can safely use assertions in the class since files won't be removed unless the recovery
is either cancelled, failed or finished.

Closes #19473

### CODE REVIEW

**Comment by s1monw** [comment]
@bleskes can you take a look?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java [hunk 1]
```diff
@@ -68,6 +69,35 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
...
+            try {
+                if (onGoingRecoveries.replace(ref.status.recoveryId(), ref.status, copy) == false) {
+                    throw new IllegalStateException("failed to reset recovery");
+                }
+                ref.status.close();
+                copy.indexShard().performRecoveryRestart();
+                success = true;
+            } finally {
+                if (success == false) {
+                    copy.cancel("reset failed");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java]: we need to remove this from onGoingRecoveries and only call cancel if it was found, otherwise we may leave a lingering recovery id, pointing at a cancelled recovery.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 2]
```diff
@@ -151,6 +160,13 @@ public void renameAllTempFiles() throws IOException {
         store.renameTempFilesSafe(tempFileNames);
     }
 
+    public void close() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: I'm a bit concerned with exposing this close() as public is - someone may think this is the way to finish / clean up (which you should call `cancel` or `markAsDone`). How about calling this `closeAndStartNew` and make it return the new RecoveryTarget ? this also means we can make the copy constructor private, which is a good thing?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 3]
```diff
@@ -78,27 +78,36 @@
...
         super("recovery_status");
-        this.recoveryId = idGenerator.incrementAndGet();
+        this.cancellableThreads = cancellableThreads;
+        this.recoveryId = recoveryId;
         this.listener = listener;
         this.logger = Loggers.getLogger(getClass(), indexShard.indexSettings().getSettings(), indexShard.shardId());
         this.indexShard = indexShard;
         this.sourceNode = sourceNode;
         this.shardId = indexShard.shardId();
         this.tempFilePrefix = RECOVERY_PREFIX + indexShard.recoveryState().getTimer().startTime() + ".";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: hmm.. wondering if we should add a unique generation to these files (or sample the idGenerator again when we restart). Right now we will have collisions between temp names of the previous recovery target and the new one, but I think we want complete separation?

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 4]
```diff
@@ -197,7 +201,7 @@ protected ReplicationGroup createGroup(int replicas) throws IOException {
         return new ReplicationGroup(metaData, homePath);
     }
 
-    private DiscoveryNode getDiscoveryNode(String id) {
+    protected DiscoveryNode getDiscoveryNode(String id) {
         return new DiscoveryNode(id, id, LocalTransportAddress.buildUnique(), Collections.emptyMap(),
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: wondering if we should switch `LocalTransportAddress.buildUnique()` to `new LocalTransportAddress(id)`

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 5]
```diff
@@ -227,7 +231,7 @@ private IndexShard newShard(boolean primary, DiscoveryNode node, IndexMetaData i
     }
 
 
-    class ReplicationGroup implements AutoCloseable, Iterable<IndexShard> {
+    public class ReplicationGroup implements AutoCloseable, Iterable<IndexShard> {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: why public? protected not good enough?

File: core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java [hunk 6]
```diff
@@ -43,7 +50,7 @@
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.lessThan;
 
-public class RecoveriesCollectionTests extends ESSingleNodeTestCase {
+public class RecoveriesCollectionTests extends ESIndexLevelReplicationTestCase {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: yay

File: core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java [hunk 7]
```diff
@@ -57,83 +64,112 @@ public void onRecoveryFailure(RecoveryState state, RecoveryFailedException e, bo
...
     public void testLastAccessTimeUpdate() throws Exception {
-        createIndex();
-        final RecoveriesCollection collection = new RecoveriesCollection(logger, getInstanceFromNode(ThreadPool.class));
-        final long recoveryId = startRecovery(collection);
-        try (RecoveriesCollection.RecoveryRef status = collection.getRecovery(recoveryId)) {
-            final long lastSeenTime = status.status().lastAccessTime();
-            assertBusy(new Runnable() {
-                @Override
-                public void run() {
+        try (ReplicationGroup shards = createGroup(1)) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: I think we want createGroup(0)?

File: core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java [hunk 8]
```diff
@@ -57,83 +64,112 @@ public void onRecoveryFailure(RecoveryState state, RecoveryFailedException e, bo
...
 
-    long startRecovery(RecoveriesCollection collection, RecoveryTargetService.RecoveryListener listener, TimeValue timeValue) {
-        IndicesService indexServices = getInstanceFromNode(IndicesService.class);
-        IndexShard indexShard = indexServices.indexServiceSafe(resolveIndex("test")).getShardOrNull(0);
-        final DiscoveryNode sourceNode = new DiscoveryNode("id", LocalTransportAddress.buildUnique(), emptyMap(), emptySet(),
-            Version.CURRENT);
+    long startRecovery(RecoveriesCollection collection, DiscoveryNode sourceNode, IndexShard indexShard,
+                       RecoveryTargetService.RecoveryListener listener, TimeValue timeValue) {
+        final DiscoveryNode rNode = getDiscoveryNode(indexShard.routingEntry().currentNodeId());
+        indexShard.markAsRecovering("remote", new RecoveryState(indexShard.shardId(), false, RecoveryState.Type.REPLICA, sourceNode,
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: I spent some time thinking how we can fold this into the ReplicationGroup in a clean way. I couldn't come up with anything fantastic - the best way I had was to make the actually recovery code in ReplicationGroup be a method on it's own which these tests can override and only do the recovery collection part. not sure how much cleaner that will be.

**Comment by bleskes** [comment]
It took a while to convince my self that keeping the same recovery id is the right way to go for now. I like how it's done and left some minor comments. Thx @s1monw

File: core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java (hunk 8 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: I think we can just stick with it once we need it again, refactor as we go..

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: agreed

**Comment by s1monw** [comment]
@bleskes I pushed new commits

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java [hunk 9]
```diff
@@ -77,22 +78,22 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
...
+                if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
+                    throw new IllegalStateException("failed to replace recovery target");
                 }
-                ref.status.close();
-                copy.indexShard().performRecoveryRestart();
                 success = true;
             } finally {
                 if (success == false) {
-                    copy.cancel("reset failed");
+                    onGoingRecoveries.remove(id);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java]: nit: now that we folded the close and performRecoveryRestart into `status.resetRecovery()` we don't need the success pattern anymore. This can be:

```
if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
       resetRecovery.cancel("replace failed");
       throw new IllegalStateException("failed to replace recovery target");
}

```

File: core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java [hunk 10]
```diff
@@ -146,6 +147,8 @@ public void testResetRecovery() throws Exception {
                     recovery.close();
                     expectThrows(ElasticsearchException.class, () -> recovery.status().store());
                     assertEquals(referencesToStore, resetRecovery.status().store().refCount());
+                    String resetTempFileName = resetRecovery.status().getTempNameForFile("foobar");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java]: ++

File: core/src/test/java/org/elasticsearch/indices/recovery/RecoveryStatusTests.java [hunk 11]
```diff
@@ -73,7 +71,7 @@ public void onRecoveryFailure(RecoveryState state, RecoveryFailedException e, bo
         Set<String> strings = Sets.newHashSet(status.store().directory().listAll());
         String expectedFile = null;
         for (String file : strings) {
-            if (Pattern.compile("recovery[.]\\d+[.]foo[.]bar").matcher(file).matches()) {
+            if (Pattern.compile("recovery[.][\\w-]+[.]foo[.]bar").matcher(file).matches()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/recovery/RecoveryStatusTests.java]: :)

**Comment by bleskes** [comment]
LGTM. Left minor suggestions - feel free to accept or reject and push away.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java [hunk 12]
```diff
@@ -68,6 +70,35 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
...
+            RecoveryTarget resetRecovery = status.resetRecovery();
+            boolean success = false;
+            try {
+                if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
+                    throw new IllegalStateException("failed to replace recovery target");
+                }
+                success = true;
+            } finally {
+                if (success == false) {
+                    onGoingRecoveries.remove(id);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java]: copying from a comment on the commit- 

nit: now that we folded the close and performRecoveryRestart into `status.resetRecovery()` we don't need the success pattern anymore. This can be:

```
if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
       resetRecovery.cancel("replace failed");
       throw new IllegalStateException("failed to replace recovery target");
}

```
